x/programs/simulator/state/state.go

```diff
@@ -0,0 +1,172 @@
+// Copyright (C) 2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package state
+
+/*
+#cgo CFLAGS: -I../common
+#include "callbacks.h"
+#include <stdlib.h>
+*/
+import "C"
+
+import (
+	"context"
+	"fmt"
+	"unsafe"
+
+	"github.com/ava-labs/avalanchego/database"
+
+	"github.com/ava-labs/hypersdk/codec"
+	"github.com/ava-labs/hypersdk/state"
+)
+
+// ensure SimulatorState implements state.Mutable
+var _ state.Mutable = &SimulatorState{}
+
+type SimulatorState struct {
+	// this is a pointer to the state passed in from rust
+	// it points to the state object on the rust side of the simulator
+	statePtr unsafe.Pointer
+
+	// this is ptr to the get function
+	getFunc C.GetStateCallback
+
+	// this is ptr to the insert function
+	insertFunc C.InsertStateCallback
+
+	// this is a ptr to the delete function
+	removeFunc C.RemoveStateCallback
+}
+
+func NewSimulatorState(db unsafe.Pointer) *SimulatorState {
+	// convert unsafe pointer to C.Mutable
+	mutable := (*C.Mutable)(db)
+	return &SimulatorState{
+		statePtr:   mutable.stateObj,
+		getFunc:    mutable.get_value_callback,
+		insertFunc: mutable.insert_callback,
+		removeFunc: mutable.remove_callback,
+	}
+}
+
+func (s *SimulatorState) GetValue(_ context.Context, key []byte) ([]byte, error) {
+	return s.getValue(key)
+}
+
+func (s *SimulatorState) Insert(_ context.Context, key []byte, value []byte) error {
+	return s.insert(key, value)
+}
+
+func (s *SimulatorState) Remove(_ context.Context, key []byte) error {
+	return s.remove(key)
+}
+
+// In C, a function argument written as a fixed size array actually requires a
+// pointer to the first element of the array. C compilers are aware of this calling
+// convention and adjust the call accordingly, but Go cannot. In Go, you must pass
+// the pointer to the first element explicitly: C.f(&C.x[0]).
+//
+// getFuncPtr is a pointer to the get state callback function
+// dbPtr is a pointer to the state object on the rust side of the simulator
+// key is the key to be used to get the value from the state object
+// this function returns the value grabbed from the state object
+func (s *SimulatorState) getValue(key []byte) ([]byte, error) {
+	bytesPtr := C.CBytes(key)
+	defer C.free(bytesPtr)
+
+	bytesStruct := C.Bytes{
+		data:   (*C.uint8_t)(bytesPtr),
+		length: C.uint(len(key)),
```

Comment:
We should also be able to use a 64-bit type here so that we can just define it as a `usize` in Rust... or is there no way to do that with `bindgen`?